src/source-windivert.c

```diff
@@ -0,0 +1,682 @@
+static TmEcode WinDivertRecvHelper(ThreadVars *tv, WinDivertThreadVars *);
+static TmEcode WinDivertVerdictHelper(ThreadVars *tv, Packet *p);
+static TmEcode WinDivertCloseHelper(WinDivertThreadVars *);
+
+void TmModuleReceiveWinDivertRegister(void)
+{
+    SCMutexInit(&g_wd_init_lock, NULL);
+
+    TmModule *tm_ptr = &tmm_modules[TMM_RECEIVEWINDIVERT];
+    memset(tm_ptr, 0, sizeof(TmModule));
+
+    tm_ptr->name = "ReceiveWinDivert";
+    tm_ptr->ThreadInit = ReceiveWinDivertThreadInit;
+    tm_ptr->PktAcqLoop = ReceiveWinDivertLoop;
+    tm_ptr->ThreadExitPrintStats = ReceiveWinDivertThreadExitStats;
+    tm_ptr->ThreadDeinit = ReceiveWinDivertThreadDeinit;
+    tm_ptr->flags = TM_FLAG_RECEIVE_TM;
+}
+
+void TmModuleVerdictWinDivertRegister(void)
+{
+    TmModule *tm_ptr = &tmm_modules[TMM_VERDICTWINDIVERT];
+    memset(tm_ptr, 0, sizeof(TmModule));
+
+    tm_ptr->name = "VerdictWinDivert";
+    tm_ptr->ThreadInit = VerdictWinDivertThreadInit;
+    tm_ptr->Func = VerdictWinDivert;
+    tm_ptr->ThreadDeinit = VerdictWinDivertThreadDeinit;
+}
+
+void TmModuleDecodeWinDivertRegister(void)
+{
+    TmModule *tm_ptr = &tmm_modules[TMM_DECODEWINDIVERT];
+    memset(tm_ptr, 0, sizeof(TmModule));
+
+    tm_ptr->name = "DecodeWinDivert";
+    tm_ptr->ThreadInit = DecodeWinDivertThreadInit;
+    tm_ptr->Func = DecodeWinDivert;
+    tm_ptr->ThreadDeinit = DecodeWinDivertThreadDeinit;
+    tm_ptr->flags = TM_FLAG_DECODE_TM;
+}
+
+/**
+ * \brief Main WinDivert packet receive pump
+ */
+TmEcode ReceiveWinDivertLoop(ThreadVars *tv, void *data, void *slot)
+{
+    SCEnter();
+
+    WinDivertThreadVars *wd_tv = (WinDivertThreadVars *)data;
+    wd_tv->slot = ((TmSlot *)slot)->slot_next;
+
+    while (true) {
+        if (suricata_ctl_flags & SURICATA_STOP) {
+            SCReturnInt(TM_ECODE_OK);
+        }
+
+        if (unlikely(WinDivertRecvHelper(tv, wd_tv) != TM_ECODE_OK)) {
+            SCReturnInt(TM_ECODE_FAILED);
+        }
+
+        StatsSyncCountersIfSignalled(tv);
+    }
+
+    SCReturnInt(TM_ECODE_OK);
+}
+
+static TmEcode WinDivertRecvHelper(ThreadVars *tv, WinDivertThreadVars *wd_tv)
+{
+    SCEnter();
+
+#ifdef COUNTERS
+    WinDivertQueueVars *wd_qv = WinDivertGetQueue(wd_tv->thread_num);
+#endif /* COUNTERS */
+
+    /* make sure we have at least one packet in the packet pool, to prevent us
+     * from alloc'ing packets at line rate
+     */
+    PacketPoolWait();
+
+    /* obtain a packet buffer */
+    Packet *p = PacketGetFromQueueOrAlloc();
+    if (unlikely(p == NULL)) {
+        SCLogDebug(
+                "PacketGetFromQueueOrAlloc() - failed to obtain Packet buffer");
+        SCReturnInt(TM_ECODE_FAILED);
+    }
+    PKT_SET_SRC(p, PKT_SRC_WIRE);
+
+    /* WinDivert needs to be fed a buffer, so we must make one available. It is
+     * highly likely we'll encounter segmentation offload so we'll just give our
+     * pool packets external buffers.
+     */
+    PacketCallocExtPkt(p, MAX_PAYLOAD_SIZE);
```

Comment:
I don't think this is a good fit. It's not resizing ext pkt if needed. The PacketCopyData function will expand packets as needed.

Reply:
I'm definitely _way open_ to suggestions here; the way it's currently written is because `WinDivertRecv` takes a buffer and writes to it, rather than providing a buffer to use/copy.

I could probably alloc an intermediate buffer per thread that then gets `PacketCopyData`'d

Reply:
I think the cleanest way to address this would be to add support for your packet size for ConfigGetCaptureValue() in suricata.c. It uses some logic like getting the MTU to figure out what default allocation size we should use in the PacketPool.